freeipa/src/main/java/com/sequenceiq/freeipa/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -20,24 +32,92 @@
     @Inject
     private MultiAzValidator multiAzValidator;
 
-    public void calculateByRoundRobin(Network network, InstanceGroup instanceGroup) {
-        String subnetId = getSubnetId(network);
-        if (!Strings.isNullOrEmpty(subnetId) && multiAzValidator.supportedForInstanceMetadataGeneration(network.cloudPlatform())) {
+    public Map<String, String> prepareSubnetAzMap(DetailedEnvironmentResponse environment) {
+        Map<String, String> subnetAzPairs = new HashMap<>();
+        if (environment != null && environment.getNetwork() != null) {
+            for (Map.Entry<String, CloudSubnet> entry : environment.getNetwork().getSubnetMetas().entrySet()) {
+                CloudSubnet value = entry.getValue();
+                if (!Strings.isNullOrEmpty(value.getName())) {
+                    subnetAzPairs.put(value.getName(), value.getAvailabilityZone());
+                }
+                if (!Strings.isNullOrEmpty(value.getId())) {
+                    subnetAzPairs.put(value.getId(), value.getAvailabilityZone());
+                }
+            }
+        }
+        return subnetAzPairs;
+    }
+
+    public void calculateByRoundRobin(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup, String variant) {
+        Map<String, Integer> subnetUsage = new HashMap<>();
+        Set<String> subnetIds = collectSubnetIds(instanceGroup);
+        initializeSubnetUsage(subnetAzPairs, subnetIds, subnetUsage);
+        collectCurrentSubnetUsage(instanceGroup, subnetUsage);
+        if (!subnetIds.isEmpty() && multiAzValidator.supportedForInstanceMetadataGeneration(instanceGroup)) {
             for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
```

Comment:
After a bit of thinking, I am not sure anymore if this will make any difference. In `core` it was essential for sake of upscale & repair to be able to cope with instances that got terminated / deleted previously (either as part of a downscale, or manual terminatation on the provider side). I wonder what the repair story is for FreeIPA.